### ISSUE TITLE
Introduce a 2d array data structure for buffering

### ISSUE THREAD

**Issue by mstimberg** [brian-team/brian2#467]
In various places, most importantly for brian hears, we will need buffer for analog values. To support this, we should add a specific data structure. The data structure itself will be straightforward, a circular array that allows to insert new values and to retrieve previously stored values.

One main use case (apart from brian hears) is to enable delayed continuous connections between neurons, e.g. for rate models. The question is then how this functionality should be exposed to the user. I can think of at least two ways:

* extend `linked_var` for this purpose:
```Python
G1 = NeuronGroup(..., 'dv/dt = ... ' )
G2 = NeuronGroup(..., 'v_delayed : 1 (linked)')
G2.v_delayed = linked_var(G1, 'v', delays=1*ms)
```
and for full flexibility, we'd also allow something like `linked_var(G1, 'v', delays='delay', max_delay=5*ms)`, referring to a state variable `delay`.

* introduce a class similar to `TimedArray`:
```Python
G1 = NeuronGroup(..., 'dv/dt = ... ' )
v_delayed = BufferedCopy(G1, 'v', buffer=1*ms)
G2 = NeuronGroup(..., 'dv/dt = ... v_delayed(1*ms, i) ...')
```

**Comment by mstimberg** [comment]
We briefly discussed this in our meeting today, there was a consensus that something similar to `TimedArray` is the best, in particular because it gives full flexibility (the delay can differ between neurons, change over time, etc.). It should use "absolute time", however:
```Python
G2 = NeuronGroup(..., 'dv/dt = ... v_delayed(t - 1*ms, i) ...')
```
This way, equations will be clearer and look more similar to the equations you see in papers for rate-based models, e.g. something like `dr/dt = (-r + inp(t - delta)) / tau`

**Comment by thesamovar** [comment]
I like the syntax for rate based with delays, but does the same reasoning apply for Brian hears? I might split the proposal into two parts: a 2d array for buffering AND a syntax for referring to variables with delays (that would use the buffered 2d array behind the scenes).

There's another issue: what happens if you request a time outside the recorded interval? Perhaps it should raise a RuntimeError?

**Comment by mstimberg** [comment]
> I like the syntax for rate based with delays, but does the same reasoning apply for Brian hears? I might split the proposal into two parts: a 2d array for buffering AND a syntax for referring to variables with delays (that would use the buffered 2d array behind the scenes).

I don't think we necessary need a new issue for this, I think the data structure is clear (basically a simpler version of the spike queue -- probably the spike queue can use the new data structure internally) and does not need much discussion. From my side, we don't really have to decide what to do for brian hears right now, maybe some lower-level code will use the data-structure directly while some higher-level code will use the above syntax?

> There's another issue: what happens if you request a time outside the recorded interval? Perhaps it should raise a RuntimeError?

This is a tricky question, we currently don't do any such error checking during runtime (e.g. in `TimedArray` you can provide a non-existing neuron index). Maybe open a new issue for this? I could think of a global "debug" flag that would switch on checks, for example.

**Comment by thesamovar** [comment]
OK, yes, let's just make sure that the data structure is not tied to the syntax, indeed.

Good idea about runtime error checking. It seems that it's getting more and more important. I'll make a new issue for that now and we can drop it from the discussion here.
